### PR TITLE
OpenMPTarget: MDRange parallel_reduce

### PR DESCRIPTION
**Pull request by rgayatri23**
The PR contains the following:

1. MDRange parallel_reduce implementation
2. Implement copy-back in cases where `reduction` happens on a device view.
3. Enable corresponding tests.

### CODE REVIEW

File: core/src/Kokkos_OpenMPTargetSpace.hpp [hunk 1]
```diff
@@ -51,13 +51,12 @@
...
 
 #include <Kokkos_Core_fwd.hpp>
-
-#ifdef KOKKOS_ENABLE_OPENMPTARGET
-
 #include <OpenMPTarget/Kokkos_OpenMPTarget_Error.hpp>
 #include <Kokkos_HostSpace.hpp>
 #include <omp.h>
 
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
```
Comment (dalg24) [inline core/src/Kokkos_OpenMPTargetSpace.hpp:58]: Why did you change?  Don't we need to protect the include of `<omp.h>`?

Comment (rgayatri23) [reply core/src/Kokkos_OpenMPTargetSpace.hpp]: Right, dint notice that, <omp.h> came out of the `ifdef` . Fixed it.

**Comment by masterleinad** [comment]
```
/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_Profiling.hpp:259:20: error: no member named 'max_tile_size_product' in 'Kokkos::Impl::ParallelReduce<(lambda at /var/jenkins/workspace/Kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9), Kokkos::MDRangePolicy<Kokkos::Experimental::OpenMPTarget, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::InvalidType, Kokkos::Experimental::OpenMPTarget>'
    return driver::max_tile_size_product(policy, functor);
                   ^
```

**Comment by rgayatri23** [comment]
> ```
> /var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_Profiling.hpp:259:20: error: no member named 'max_tile_size_product' in 'Kokkos::Impl::ParallelReduce<(lambda at /var/jenkins/workspace/Kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9), Kokkos::MDRangePolicy<Kokkos::Experimental::OpenMPTarget, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::InvalidType, Kokkos::Experimental::OpenMPTarget>'
>     return driver::max_tile_size_product(policy, functor);
>                    ^
> ```

Fixed it.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp [hunk 2]
```diff
@@ -810,35 +698,499 @@ class ParallelReduce<FunctorType, Kokkos::MDRangePolicy<Traits...>, ReducerType,
...
+    if (m_result_ptr_on_device)
+      OMPT_SAFE_CALL(omp_target_memcpy(ptr, &result, sizeof(double), 0, 0,
+                                       omp_get_default_device(),
+                                       omp_get_initial_device()));
+    else
+      *ptr = result;
+  }
+
+  template <int Rank, class ValueType>
+  inline typename std::enable_if<Rank == 7>::type execute_tile(
```
Comment (dalg24) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp:1103]: We don't support `rank == 7`
https://github.com/kokkos/kokkos/blob/8f33b3873e2f26a37eaf398fb876d0804a8f369a/core/src/Kokkos_Rank.hpp#L60

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp [hunk 3]
```diff
@@ -810,35 +698,499 @@ class ParallelReduce<FunctorType, Kokkos::MDRangePolicy<Traits...>, ReducerType,
...
-    //                                , Kokkos::HostSpace >::value
-    //  , "Reduction result on Kokkos::Experimental::OpenMPTarget must be a
-    //  Kokkos::View in HostSpace" );
+        m_result_ptr_on_device(
+            MemorySpaceAccess<Kokkos::Experimental::OpenMPTargetSpace,
+                              typename ReducerType::result_view_type::
+                                  memory_space>::accessible) {}
+
+  template <int Rank, class ValueType>
+  inline typename std::enable_if<Rank == 1>::type execute_tile(
```
Comment (dalg24) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp:721]: `Rank` has to be at least 2

**Review by dalg24 — CHANGES_REQUESTED** [review]
Looks fine besides the invalid specializations for rank 1 and 7

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp (hunk 3 above)
Comment (rgayatri23) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp]: 👍  Delete implementations for rank > 6 and rank < 2.

**Review by dalg24 — APPROVED** [review]
